### PR TITLE
add usage examples for conflate / conflate with seed

### PR DESCRIPTION
**Pull request by ktoso**
Refs https://github.com/akka/akka/issues/25468

Adds examples for conflate and conflate with seed.

### CODE REVIEW

File: akka-docs/src/test/scala/docs/stream/IntegrationDocSpec.scala
Comment (raboof) [inline akka-docs/src/test/scala/docs/stream/IntegrationDocSpec.scala]: Perhaps also add the includes to the snippets?

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md
Comment (raboof) [inline akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md]: I see we have the 'semantics' section after the example in other places as well, but wouldn't it make sense to have it above?

**Review by raboof — COMMENTED** [review]
Would it be nice to have the example in 'production code form' rather than 'unit test form'? This is an improvement to the current docs though in any case ;).

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md
Comment (ktoso) [reply akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md]: My argument is that they're less helpful, people jump here to get an example, then can dig in deeper

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md
Comment (patriknw) [inline akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md]: should we standardize on where to put those snippets for the operators docs, such as in existing https://github.com/akka/akka/tree/master/akka-docs/src/test/scala/docs/stream/operators

To make it easy for contributors of those examples (because the core team will not have time to write them all) we should keep the ambition rather low. They don't have to be full tests. Functionality should be tested elsewhere.

Also, IntegrationDocSpec is about integration with actors (or outside world) so I don't think that is a good place for this one

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md
Comment (patriknw) [reply akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md]: whatever, consistency is king, so then change all

File: akka-docs/src/test/java/jdocs/stream/IntegrationDocTest.java
Comment (patriknw) [inline akka-docs/src/test/java/jdocs/stream/IntegrationDocTest.java]: I think those snippets should be small and to the point without showing testing stuff. I can see that in here you want to show how the result differ with demand. Perhaps then first have a very concise one-liner of the `conflate` and then the full example with some additional explanation.

**Review by patriknw — COMMENTED** [review]
Some comments because I would like that the first additions of these act as good examples what we expect from contributors of such examples.

File: akka-docs/src/test/java/jdocs/stream/IntegrationDocTest.java
Comment (ktoso) [reply akka-docs/src/test/java/jdocs/stream/IntegrationDocTest.java]: Ok makes sense, will do

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md
Comment (ktoso) [reply akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md]: I see; the location for the example was based on a different operator docs; that one was not following this style as well then. I'll align them all together in this PR

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md
Comment (ktoso) [reply akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md]: Keeping the consistent one then; as-is

**Comment by ktoso** [comment]
Polishing this up now...

**Comment by ktoso** [comment]
I made the snippets be very simple rather than the crazy intense tests.

File: akka-docs/src/test/java/jdocs/stream/operators/SourceOrFlow.java
Comment (patriknw) [inline akka-docs/src/test/java/jdocs/stream/operators/SourceOrFlow.java]: those imports are not part of the log example

File: akka-docs/src/test/java/jdocs/stream/operators/SourceOrFlow.java
Comment (patriknw) [inline akka-docs/src/test/java/jdocs/stream/operators/SourceOrFlow.java]: strange indentation here?

**Review by patriknw — APPROVED** [review]
LGTM, with minor nitpicks

File: akka-docs/src/test/java/jdocs/stream/operators/SourceOrFlow.java
Comment (ktoso) [reply akka-docs/src/test/java/jdocs/stream/operators/SourceOrFlow.java]: good catch, thanks

**Comment by ktoso** [comment]
Fixed indent and imports.

Ok to merge or please someone else review @johanandren @jrudolph @raboof

**Review by jrudolph — COMMENTED** [review]
Hmm, the examples show the basic syntax but `conflate` isn't useful in a synchronous setting (it's not only not useful but it does surprising things in a synchronous setting because a synchronous upstream can actually starve the downstream side of the stream up to the size of the event horizon).

Maybe showing an example that uses `delay` or `throttle` on the upstream side and `throttle` on the downstream side and then showing the result at the end would make that clearer?

**Comment by ktoso** [comment]
The text is there to explain the semantics though.

I'll add a throttle after the `conflateWithSeed` if you think this will help explaining it.

**Comment by jrudolph** [comment]
> I'll add a `throttle` after the `conflateWithSeed` if you think this will help explaining it.

The text doesn't say anything about that the upstream must be async for `conflate` to make sense.

That's also why a `throttle` *after* `conflate` doesn't help. The important thing is that the upstream must be asynchronous for `conflate` to make sense.

**Comment by ktoso** [comment]
It talks about upstream being "slower" which means they cant be fused... but that is indeed a great catch -- the actor based tests were async by their normal nature, and the throttle would be throttled into this so indeed yielding a predictable outcome :O 

Thanks @jrudolph, I'll address this by adding .async inn the right spot. I even think I made this mistake in the Coursera materials so will fix this there as well :-)

**Comment by ktoso** [comment]
So I actually tried out things to confirm my thinking... And no, we do not need to insert any asyncs there for the examples to be correct; See:

```
```

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md]: I think this could be simplified (and fix mixed tense). What about something like this: 
"If downstream is slower the elements is conflated by summing them. This means that upstream can continue producing elements while downstream is applying backpressure. For example: downstream is backpressuring while `1`, `10` and `100` arrives from upstream, then backpressure stops and the conflated `111` is emitted downstream."

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflateWithSeed.md
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflateWithSeed.md]: Skip first comma. "while there is backpressure" instead of the "upheld" phrasing"?

File: akka-docs/src/test/java/jdocs/stream/operators/SourceOrFlow.java
Comment (johanandren) [inline akka-docs/src/test/java/jdocs/stream/operators/SourceOrFlow.java]: Maybe call it `SummedValue` or something that describes what it does rather than this?

**Review by johanandren — APPROVED** [review]
Added some nitpicks and potential improvements, but LGTM enough to merge if you disagree with them.

**Comment by jrudolph** [comment]
> And no, we do not need to insert any asyncs there for the examples to be correct

Ah, maybe I wasn't clear. I didn't meant that explicit async boundaries are necessary but any kind of async *behavior* on the source side. `throttle` is enough because it uses timers or async callbacks to throttle.

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md
Comment (ktoso) [reply akka-docs/src/main/paradox/stream/operators/Source-or-Flow/conflate.md]: Thanks, applied

**Comment by ktoso** [comment]
> > And no, we do not need to insert any asyncs there for the examples to be correct
> 
> Ah, maybe I wasn't clear. I didn't meant that explicit async boundaries are necessary but any kind of async _behavior_ on the source side. `throttle` is enough because it uses timers or async callbacks to throttle.

Right that's true. 

Addressed Johan's comments as well. I'll merge this once validates

**Pull request merged**
